doc/services/rtio/index.rst

```diff
@@ -0,0 +1,417 @@
+
+Using DMA and/or interrupt driven I/O shouldn't dictate whether or not the
+call is blocking or not.
+
+Inspiration, introducing io_uring
+*********************************
+
+It's better not to reinvent the wheel (or ring in this case) and io_uring as an
+API from the Linux kernel provides a winning model. In io_uring there are two
+lock-free ring buffers acting as queues shared between the kernel and a userland
+application. One queue for submission entries which may be chained and flushed to
+create concurrent sequential requests. A second queue for completion events.
+Only a single syscall is actually required to execute many operations, the
+io_uring_submit call. This call may block the caller when a number of
+operations to wait on is given.
+
+This model maps well to DMA and interrupt driven transfers. A request to do a
+sequence of operations in an asynchronous way directly relates
+to the way hardware typically works with interrupt driven state machines
+potentially involving multiple peripheral IPs like bus and DMA controllers.
+
+Submission Queue and Chaining
+*****************************
+
+The submission queue (sq), is the description of the operations
+to perform in concurrent chains.
+
+For example imagine a typical SPI transfer where you wish to write a
+register address to then read from. So the sequence of operations might be...
+
+   1. Chip Select
+   2. Clock Enable
+   3. Write register address into SPI transmit register
+   4. Read from the SPI receive register into a buffer
+   5. Disable clock
+   6. Disable Chip Select
+
+If anything in this chain of operations fails give up. Some of those operations
+can be embodied in a device abstraction that understand a read or write
+implicitly means setup the clock and chip select. The transactional nature of
+the request also needs to be embodied in some manner. Of the operations above
+perhaps the read could be done using DMA as its large enough make sense. That
+requires an understanding of how to setup the device's particular DMA to do so.
+
+The above sequence of operations is embodied in RTIO as chain of
+submission queue entries (sqe). Chaining is done by setting a bitflag in
+an sqe to signify the next sqe must wait on the current one.
+
+Because the chip select and clocking is common to a particular SPI controller
+and device on the bus it is embodied in what RTIO calls an iodev.
+
+Multiple operations against the same iodev are done in the order provided as
+soon as possible. If two operation chains have varying points using the same
+device its possible one chain will have to wait for another to complete.
+
+Completion Queue
+****************
+
+In order to know when a sqe has completed there is a completion
+queue (cq) with completion queue events (cqe). A sqe once completed results in
+a cqe being pushed into the cq. The ordering of cqe may not be the same order of
+sqe. A chain of sqe will however ensure ordering and failure cascading.
```

Comment:
The ordering may not be the same. The cqe and sqe are tied together with the common void *userdata which is passed along. While this userdata could point at something like a struct or buffer, it could also directly store an integer identifier like a sequential index, that's up to the user of the API.

Reply:
So it is left to the originator to determine which CQE it has just consumed, correct?

Reply:
Yes it is, that's correct.

Reply:
It sort of seems like adding the offset index of the entry that is being sent to the executor would be easy enough and would allow the originator less grief finding what has and has not been executed in a common orderly manner.